napari/layers/shapes/shapes.py

```diff
@@ -684,10 +681,20 @@ def add(
 
                 # Add shape
                 self._data_view.add(shape)
-                self.face_color[shape_idx, :] = shape.face_color
-                self.edge_color[shape_idx, :] = shape.edge_color
-            self.current_face_color = shape.face_color
-            self.current_edge_color = shape.edge_color
+                new_face_colors[shape_idx, :] = shape.face_color
+                new_edge_colors[shape_idx, :] = shape.edge_color
+            self._current_face_color = shape.face_color
+            self._current_edge_color = shape.edge_color
+
+            # If we just instantiated the layer then we can directly set
+            # new face and edge colors. Else we'll append to the existing
+            # ones the new values.
+            if len(self._data_view) == num_of_shapes:
+                self._face_color = new_face_colors
+                self._edge_color = new_edge_colors
+            else:
+                self._face_color = np.concatenate((self.face_color, new_face_colors), axis=0)
+                self._edge_color = np.concatenate((self.edge_color, new_edge_colors), axis=0)
```

Comment:
> I believe that the better option is to only hold the color and opacity data in the ShapeList by modifying its existing face_color, edge_color and opacity attributes. When the Shapes layers needs this data it will simply ask the ShapeList for its colors and opacity.

This sounds good to me too - the guiding philosophy of only having one source of ground truth is a good one. Here we have this distinction between `ShapeList` and the main `Shapes` layer, that hopefully can the code more modular, but maybe introduces additional complexities like what you mentioned above.